cmd/kubeadm/app/cmd/phases/kubelet.go

```diff
@@ -0,0 +1,176 @@
+}
+
+// NewCmdKubeletWriteInitConfig calls cobra.Command for writing init kubelet configuration
+func NewCmdKubeletWriteInitConfig() *cobra.Command {
+	var cfgPath string
+	cmd := &cobra.Command{
+		Use:     "write-init-config",
+		Short:   "Write init kubelet configuration to disk",
+		Long:    kubeletWriteInitConfigLongDesc,
+		Example: kubeletWriteInitConfigExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cfg := &kubeadmapiext.MasterConfiguration{}
+			legacyscheme.Scheme.Default(cfg)
+
+			// This call returns the ready-to-use configuration based on the configuration file that might or might not exist and the default cfg populated by flags
+			internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, cfg)
+			kubeadmutil.CheckErr(err)
+			err = kubeletphase.WriteInitKubeletConfigToDiskOnMaster(internalcfg)
+			kubeadmutil.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().StringVar(&cfgPath, "config", cfgPath, "Path to kubeadm config file (WARNING: Usage of a configuration file is experimental)")
+
+	return cmd
+}
+
+// NewCmdKubeletUploadDynamicConfig calls cobra.Command for uploading dynamic kubelet configuration
+func NewCmdKubeletUploadDynamicConfig() *cobra.Command {
+	var cfgPath, kubeConfigFile string
+
+	cmd := &cobra.Command{
+		Use:     "upload-dynamic-config",
+		Short:   "Upload dynamic kubelet configuration as ConfigMap",
+		Long:    kubeletUploadDynamicConfigLongDesc,
+		Example: kubeletUploadDynamicConfigExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cfg := &kubeadmapiext.MasterConfiguration{}
+			legacyscheme.Scheme.Default(cfg)
+
+			// This call returns the ready-to-use configuration based on the configuration file that might or might not exist and the default cfg populated by flags
+			internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, cfg)
+			kubeadmutil.CheckErr(err)
+			client, err := kubeconfigutil.ClientSetFromFile(kubeConfigFile)
+			kubeadmutil.CheckErr(err)
+			err = kubeletphase.CreateBaseKubeletConfiguration(internalcfg, client)
+			kubeadmutil.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().StringVar(&cfgPath, "config", cfgPath, "Path to kubeadm config file (WARNING: Usage of a configuration file is experimental)")
+	cmd.Flags().StringVar(&kubeConfigFile, "kubeconfig", "/etc/kubernetes/admin.conf", "The KubeConfig file to use when talking to the cluster")
+
+	return cmd
+}
+
+// NewCmdKubeletEnableDynamicConfig calls cobra.Command for enabling dynamic kubelet configuration on node
+func NewCmdKubeletEnableDynamicConfig() *cobra.Command {
+	cfg := &kubeadmapiext.NodeConfiguration{}
+	legacyscheme.Scheme.Default(cfg)
+
+	var cfgPath string
+	cmd := &cobra.Command{
+		Use:     "enable-dynamic-config",
+		Short:   "Enable dynamic kubelet configuration on node",
+		Long:    kubeletEnableDynamicConfigLongDesc,
+		Example: kubeletEnableDynamicConfigExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			nodeName, err := getNodeName(cfgPath, cfg)
```

Comment:
I think that you can get the `nodeName` from the config file using `configutil.ConfigFileAndDefaultsToInternalConfig`, thus avoiding to replicate the initialisation logic for this field

Reply:
Agreed. Will do.

Reply:
I took another look at the code and found that `configutil.ConfigFileAndDefaultsToInternalConfig` only applies to `MasterConfiguration` whereas the `cfg` here is `NodeConfiguration`, so I can't use that func to get the nodeName :(